proxy/request.go

```diff
@@ -118,32 +118,120 @@ func (r *request) OnClose(_ error) {
 	} else {
 		if !r.done {
 			r.done = true
-			r.send(&message.Unavailable{ErrorMessage: "No more hosts available (cluster connection closed and request is not idempotent)"})
+			r.send(&message.ServerError{ErrorMessage: "Proxy is unable to retry non-idempotent query after connection to backend cluster closed"})
 		}
 	}
 }
 
 func (r *request) OnResult(raw *frame.RawFrame) {
 	r.mu.Lock()
+	defer r.mu.Unlock()
 	if !r.done {
-		r.done = true
-		if r.raw.Header.OpCode == primitive.OpCodePrepare && raw.Header.OpCode == primitive.OpCodeResult { // Prepared result
-			frm, err := codec.ConvertFromRawFrame(raw)
-			if err != nil {
-				r.client.proxy.logger.Error("error attempting to decode prepared result message")
-			} else if _, ok := frm.Body.Message.(*message.PreparedResult); !ok { // TODO: Use prepared type data to disambiguate idempotency
-				r.client.proxy.logger.Error("expected prepared result message, but got something else")
-			} else {
-				idempotent, err := parser.IsQueryIdempotent(r.query)
+		if raw.Header.OpCode != primitive.OpCodeError ||
+			!r.handleErrorResult(raw) { // If the error result is retried then we don't send back this response
```

Comment:
Is this recursive? If the retries are happening within this method, is it possible for a deadlock to happen on `r.mu`?

Reply:
Good thought. It's not. `OnResult()` is only called as a result of server-side response.

Reply:
Oh so only the sending of the request happens within `handleErrorResult`? The response processing occurs on another component in a separate goroutine then?

Reply:
exactly, the sending of the request and the recv'ing of the request happen on different go routines.